sycl/doc/extensions/experimental/sycl_ext_oneapi_graph.asciidoc

```diff
@@ -1011,6 +1016,146 @@ If a node containing a dynamic parameter is updated through the whole graph
+being thrown.
+
+===== Restrictions
+
+The following restrictions apply to any graph using these allocations:
+
+* The graph cannot be finalized more than once.
+* The graph cannot be used as a sub-graph in another graph.
+* Graph memory allocation nodes cannot be updated, and graphs containing these
+nodes cannot be updated via <<whole-graph-update, Whole Graph Update>>.
+
+===== Behaviour 
+
+The semantics of `async_malloc` and `async_free` within a graph differ from the
+eager SYCL usage described in the
+link:../experimental/sycl_ext_oneapi_async_memory_alloc.asciidoc[sycl_ext_oneapi_async_memory_alloc]
+extension.
+
+* Graph memory allocations are not made directly from any default or
+user-provided Memory Pool. Each graph containing alloc/free nodes maintains its
+own pool of memory from which allocations are made. The following properties of
+a default or user-provided memory pool provided in calls to
+`async_<malloc/malloc_from_pool>` will be respected for the associated graph
+allocations, all other properties will be ignored:
+
+** The allocation type specified when creating the pool with
+`usm::alloc::<host/device/shared>`, subject to the limitations in the
+<<allocation-supported-features, supported features>> section.
+
+** `property::memory_pool::read_only` - Guarantee from the user that memory is
+only being read from, the implementation may be able to optimize in this case.
+
+** `property::memory_pool::zero_init` - Physical memory will be zero-initialized
+on first allocation during the call to `command_graph::finalize()`. It will not
+be zero-initialized again during execution of the graph. If that is required by
+the application it is the responsibility of the user to add the appropriate
+commands to the graph to do this.
+
+* `node_type::async_malloc` nodes within a graph will reserve a virtual address
+(VA) range on the device and return a pointer to the start of that VA range.
+This pointer can then be used in other graph nodes until an `async_free` node is
+added for that pointer. Returned pointers are not guaranteed to be unique. An
+implementation may return the same pointer as a previous `async_malloc` node if
+that pointer was previously freed via `async_free` at that point in the graph.
+
+* `node_type::async_free` nodes within a graph indicate that a given pointer/VA
+range is no longer in use, and may be available for subsequent re-use. No
+de-allocation of physical memory on a device is done during calls to
+`async_free`.
+
+* Other nodes which use a given graph allocation must be ordered via
+dependencies such that they are ordered after the alloc node and before the free
+node for a given allocation. It is the user's responsibility to ensure that
+dependencies are correct. Using a pointer in a graph command ordered after it
+has been freed via an `async_free` node results in undefined behavior.
+
+The total amount of physical memory being used by a graph can be queried using
```

Comment:
I think we may want to have a separate query or have different return value, depending on whether the graph is modifiable or finalised. When it's modifiable, this is before any optimisations are necessarily applied, so it would likely have to be the highest potential memory footprint for the graph. But then once it's finalised and optimisations for memory re-use have been applied, it can tell you the actual memory requirement on the native graph.